### PR TITLE
Simplify importer code, cleanup and extract function

### PR DESCRIPTION
**Pull request by brybacki**
**What this PR does / why we need it**:
During the first attempt to simplify importer (PR #1987) I had to drop one commit
because of some hard to find errors. This is a second attempt, and I have found
that there is an implicit communication through termination message file.

Importer writes the file and data processor appends some data there
after it is closed. This might need some future improvements.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
WDYT? Is it better?

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: cmd/cdi-importer/importer.go [hunk 1]
```diff
@@ -121,51 +106,90 @@ func main() {
...
+	volumeMode v1.PersistentVolumeMode,
+	imageSize string,
+	filesystemOverhead float64,
+	preallocation bool) int {
+	klog.V(1).Infoln("begin import process")
 
-		ds = newDataSource(source, contentType, volumeMode)
-		defer ds.Close()
+	ds := newDataSource(source, contentType, volumeMode)
+	defer func() {
```
Comment (ShellyKa13) [inline cmd/cdi-importer/importer.go:147]: Why did you put the close inside a func and not like before defer ds.Close()?

File: cmd/cdi-importer/importer.go [hunk 2]
```diff
@@ -121,51 +106,90 @@ func main() {
...
+	processor := newDataProcessor(contentType, volumeMode, ds, imageSize, filesystemOverhead, preallocation)
+	waitForReadyFile()
+	err := processor.ProcessData()
 
+	if err != nil {
+		klog.Errorf("%+v", err)
+		if err == importer.ErrRequiresScratchSpace {
+			return common.ScratchSpaceNeededExitCode
+		}
+		err = util.WriteTerminationMessage(fmt.Sprintf("Unable to process data: %+v", err))
```
Comment (ShellyKa13) [inline cmd/cdi-importer/importer.go:160]: please don't remove the err.Error(), I added it it fixes an issue: https://github.com/kubevirt/containerized-data-importer/pull/2111

File: cmd/cdi-importer/importer.go [hunk 3]
```diff
@@ -121,51 +106,90 @@ func main() {
...
-		touchDoneFile()
-		preallocationApplied = processor.PreallocationApplied()
+
+		return 1
 	}
+	touchDoneFile()
+	// due to the way some data sources can add additional information to termination message
+	// after finished (ds.close() ) termination message has to be written first, before the
+	// the ds is closed
+	// TODO: think about making communication explicit, probably DS interface should be extended
```
Comment (ShellyKa13) [inline cmd/cdi-importer/importer.go:171]: I dont completly understand if what you did in this commit fixes this issue or not, cause if so I dont see how this changes the behavior than the previous code

**Review by ShellyKa13 — COMMENTED** [review]
I think its a good clean. some question and one important comment

File: cmd/cdi-importer/importer.go (hunk 1 above)
Comment (brybacki) [reply cmd/cdi-importer/importer.go:147]: Only this part opens dataSource, so it should also close it. By doing it here I am sure that all returns from the handleImport are covered. 

Of course while doing that, I have discovered that util.WriteTerminationMessage MUST happen before ds.close().

Comment (brybacki) [reply cmd/cdi-importer/importer.go:147]: Oh, now I see, you are not asking why I did the whole thing. But why there is func wrapper?
Just a leftover from tests and debugging. Will remove now.

File: cmd/cdi-importer/importer.go (hunk 2 above)
Comment (brybacki) [reply cmd/cdi-importer/importer.go:160]: Sorry, I did play with the format, temporary added some more verbose logging. so this was removed by mistake, reverting that part

File: cmd/cdi-importer/importer.go (hunk 3 above)
Comment (brybacki) [reply cmd/cdi-importer/importer.go:171]: Let me start from beginning.

I wanted to make some small cleanup the main() so while reading the code one can see the decisions and branches (as it shows there are only two: emptyDisk, and import). 

Then I have exported methods (example: handleImport) and with that exported method I made all local variables really local. That resulted in problems. And that is when I have discovered that we have to first write Import Complete, and the we can close DataSource. 

I think this is designed in a bad way. And I may think about improving that. Some options:
1. we could add a ds.summary() or ds.terminationMessgae(). So then we just do 
```
ds.close() 
additionalInfo:= ds.terminationMessage()
err = importCompleteTerminationMessage(processor.PreallocationApplied(), additionalInfo)
```

2. made ds.close() return termination message?

Anyway, current code works exactly the same way as old. All I managed to do is extract 2 methods (handleempty and handleimport) and make all local variables really local (so this is basically last commit that I needed to drop on https://github.com/kubevirt/containerized-data-importer/pull/1987). And I have discovered that DS and importer are both trying to write to termnation message file. So this is why I have left the comment. So we can see it clearly and do not forget to improve.

File: cmd/cdi-importer/importer.go (hunk 2 above)
Comment (brybacki) [reply cmd/cdi-importer/importer.go]: done

File: cmd/cdi-importer/importer.go (hunk 1 above)
Comment (brybacki) [reply cmd/cdi-importer/importer.go]: done

File: cmd/cdi-importer/importer.go (hunk 3 above)
Comment (ShellyKa13) [reply cmd/cdi-importer/importer.go:169]: Thanks for clarifying, getting the termination message after closing sounds good to me.

**Comment by brybacki** [comment]
/retest

failed in aftersuite

**Pull request merged**